tools/mapper/find_proxy_function.py

```diff
@@ -0,0 +1,252 @@
+            min_error_index = -1
+            for i in range(0, len(proxy_list)):
+                error = get_error(trace_functions[function_name], proxy_list[i])
+                if error < min_error:
+                    min_error = error
+                    min_error_index = i
+
+            if min_error == math.inf:
+                log.warning(f"Proxy function for function {function_name} not found")
+                continue
+
+            trace_functions[function_name]["proxy-function"] = proxy_list[
+                min_error_index
+            ]["name"]
+            trace_functions[function_name]["proxy-correlation"] = get_error(
+                trace_functions[function_name], proxy_list[min_error_index]
+            )
+            log.debug(
+                f"Found proxy function for {function_name}: {trace_functions[function_name]['proxy-function']} with correlation: {trace_functions[function_name]['proxy-correlation']}"
+            )
+
+        for function_name in proxy_functions:
+            del proxy_functions[function_name]["index"]
+
+        return trace_functions, 0
+
+    except Exception as e:
+        log.error(f"Finding closest proxy function failed. Error: {e}")
+        return trace_functions, -1
+
+
+def get_proxy_function(
+    trace_functions: dict, proxy_functions: dict, unique_assignment: bool
+) -> Tuple[dict, int]:
+    """
+    Obtains the closest proxy function for every trace function
+    
+    Parameters:
+    - `trace_functions` (dict): Dictionary containing information regarding trace functions
+    - `proxy_functions` (dict): Dictionary containing information regarding proxy functions
+    - `unique_assignment` (bool): If `True`, then trace-proxy function mapping is one-to-one, provided #(proxy functions) > #(trace functions)
+    
+    Returns:
+    - `dict`: Dictionary containing information regarding trace functions with the associated proxy functions
+    - `int`: 0 if no error. -1 if error
+    """
+
+    trace_functions = OrderedDict(trace_functions)
+    proxy_functions = OrderedDict(proxy_functions)
+
+    log.info(
+        f"Lower the correlation value, the proxy function is a better proxy of the trace function"
+    )
+
+    if (unique_assignment) and (len(trace_functions) <= len(proxy_functions)):
+        log.info(
+            f"Getting One-To-One mapping between trace function and proxy function using Linear-Sum-Assignment"
+        )
+        trace_functions, err = get_proxy_function_using_linear_sum_assignment(
+            trace_functions=trace_functions, proxy_functions=proxy_functions
+        )
+        if err == -1:
+            log.error(
+                f"One-To-One mapping between trace function and proxy function not obtained"
+            )
+            log.info(
+                f"Getting closest proxy function for every trace function. Note: Mapping may not be unique"
+            )
+            trace_functions, err = get_closest_proxy_function(
```

Comment:
That is extremely unexpected behavior. If something goes sideways in the unique assignment, I'll get a non-unique one. I would be pretty annoyed to find that my tool tried to be too smart about the operation that I asked it to do. If operation is not possible, the tool should just exit.